device/sycl/include/traccc/sycl/clusterization/clusterization_algorithm.hpp

```diff
@@ -29,9 +32,9 @@ class clusterization_algorithm
     public:
     /// Constructor for clusterization algorithm
     ///
-    /// @param mr is the memory resource
+    /// @param mr is a struct of memory resources (shared or host & device)
     /// @param queue is the sycl queue for kernel invocation
-    clusterization_algorithm(vecmem::memory_resource& mr, queue_wrapper queue);
+    clusterization_algorithm(traccc::memory_resource& mr, queue_wrapper queue);
```

Comment:
This is where things become more complicated with const-correctness... :stuck_out_tongue:

We shouldn't pass the object like this. Because this formalism requires us to give an lvalue object to the algorithm. But in some situations it may be nice to be able to create the `traccc::memory_resource` object (using lvalue `vecmem::memory_resource` objects) "on the fly". Like:

```c++
vecmem::memory_resource& host_mr = ...;
vecmem::memory_resource& device_mr = ...;
sycl::queue& queue = ...;
...
traccc::clusterization_algorithm alg({host_mr, &device_mr}, &queue);
```

For the `vecmem::memory_resource` argument we actually wanted to require the user passing in an lvalue. :wink: But here we don't.

And now comes the weird const-correctness aspect of it all: You should just receive the `traccc::memory_resource` object as a const reference. After you initialise the `m_mr` member variable from it, it will not matter whether the argument to the constructor was const or not. (You could also just make the constructor receive `traccc::memory_resource` by value like we do for `traccc::sycl::queue_wrapper`. But a const reference is probably just ever slightly better to use.)

Reply:
I see, makes sense :+1: